jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java

```diff
@@ -0,0 +1,525 @@
+            return inputStream;
+        }
+        catch (IOException e)
+        {
+            throw new RuntimeIOException(e);
+        }
+    }
+
+    /**
+     * <p>Performs an asynchronous copy of the contents of a resource to a sink, using the given buffer pool and buffer characteristics.</p>
+     * <p>The resource must not be a directory, must exists and there must be
+     * a way to access its contents.</p>
+     * <p>Multiple optimized methods are used to access the resource's contents but if they all fail,
+     * {@link Resource#newInputStream()} is used as a fallback.</p>
+     *
+     * @param resource the resource to copy from.
+     * @param sink the sink to copy to.
+     * @param bufferPool the {@link ByteBufferPool} to get buffers from. null means allocate new buffers as needed.
+     * @param bufferSize the size of the buffer to be used for the copy. Any value &lt; 1 means use a default value.
+     * @param direct the directness of the buffers, this parameter is ignored if {@code bufferSize} is &lt; 1.
+     * @param callback the callback to notify when the copy is done.
+     * @throws IllegalArgumentException if the resource is a directory or does not exist or there is no way to access its contents.
+     */
+    public static void copy(Resource resource, Content.Sink sink, ByteBufferPool bufferPool, int bufferSize, boolean direct, Callback callback) throws IllegalArgumentException
+    {
+        if (resource.isDirectory() || !resource.exists())
+            throw new IllegalArgumentException("Resource must exist and cannot be a directory: " + resource);
+
+        // Save a Content.Source allocation for resources with a Path.
+        Path path = resource.getPath();
+        if (path != null)
+        {
+            try
+            {
+                new PathToSinkCopier(path, sink, bufferPool, bufferSize, direct, callback).iterate();
+            }
+            catch (Throwable x)
+            {
+                callback.failed(x);
+            }
+            return;
+        }
+
+        // Directly write the byte array if the resource is a MemoryResource.
+        if (resource instanceof MemoryResource memoryResource)
+        {
+            byte[] bytes = memoryResource.getBytes();
+            sink.write(true, ByteBuffer.wrap(bytes), callback);
+            return;
+        }
+
+        // Fallback to Content.Source.
+        Content.Source source = asContentSource(resource, bufferPool, bufferSize, direct);
+        Content.copy(source, sink, callback);
+    }
+
+    /**
+     * <p>Performs an asynchronous copy of a subset of the contents of a resource to a sink, using the given buffer pool and buffer characteristics.</p>
+     * <p>The resource must not be a directory, must exists and there must be
+     * a way to access its contents.</p>
+     * <p>Multiple optimized methods are used to access the resource's contents but if they all fail,
+     * {@link Resource#newInputStream()} is used as a fallback.</p>
```

Comment:
Isn't `ContentSource` the fallback? I don't see any calls to `Resource.newInputStream`.

Reply:
`Resource.newInputStream` is the fallback of the fallback, but you're right: `asContentSource()` is what should be documented as the fallback.